packages/node-patches/src/fs.ts

```diff
@@ -0,0 +1,345 @@
+import { Stats, Dirent, Dir } from 'fs';
+import * as path from 'path';
+import * as util from 'util';
+// using require here on purpose so we can override methods with any
+// also even though imports are mutable in typescript the cognitive dissonance is too high because es modules
+const _fs = require('fs');
+
+//tslint:disable-next-line:no-any
+export const patcher = (fs: any, root: string) => {
+  fs = fs || _fs;
+  root = root || process.env.BAZEL_SYMLINK_PATCHER_ROOT || '';
+  if (root) root = fs.realpathSync(root);
+
+  const origRealpath = fs.realpath.bind(fs);
+  const origLstat = fs.lstat.bind(fs);
+  const origReadlink = fs.readlink.bind(fs);
+  const origLstatSync = fs.lstatSync.bind(fs);
+  const origRealpathSync = fs.realpathSync.bind(fs);
+  const origReadlinkSync = fs.readlinkSync.bind(fs);
+  const origReaddir = fs.readdir.bind(fs);
+  const origReaddirSync = fs.readdirSync.bind(fs);
+
+  function isOutPath(str: string) {
+    return !root || (!str.startsWith(root + path.sep) && str !== root);
+  }
+
+  //tslint:disable-next-line:no-any
+  fs.lstat = (...args: any[]) => {
+    const cb = args.length > 1 ? args[args.length - 1] : undefined;
+    // preserve error when calling function without required callback.
+    if (cb) {
+      args[args.length - 1] = (err: Error, stats: Stats) => {
+        if (err) return cb(err);
+        if (stats.isSymbolicLink()) {
+          if (root) {
```

Comment:
This seems to imply that if `root` is `false`/`''`, then `cb` is never called..? As a meta note: Would it be possible to skip wrapping if `root` isn't provided?